README.md
# screeps-test-server
Screeps Private Server in a custom test environment. Suitable for unit/integration testing Screeps Server Mod configurations and user scripts.

---

Wrapper around Screeps Private Server. Launches a real private server using screeps/launcher. Launches a custom screeps/common process and connects to the storage process.

Manages a test server environment. Copies in a clean initial db.json, assets directory, links in a STEAM_API_KEY if one is provided, uses mods.json if one is provided.

Returns a handle to the custom common/storage process for use with Pubsub, Env, and Db direct mutations.

## Actions

`yarn build` - Builds the package, emitting .js and .d.ts files\
`yarn lint` - Runs lint over the project source\
`yarn test` - Runs all tests under the src/ directory\
`yarn publish` - Bumps package version and publishes the package to Github Packages

## Toolchain

Uses [@brisberg/typescript-pkg](https://github.com/brisberg/typescript-pkg) as a template for Toolchain configuration.

See that repo for a list of tools, documentation, and upgrade steps.


CHANGELOG.md
# Changelog
All notable changes to screeps-test-server will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.1] - 2020-06-23
### Added
- Correctly exporting typings .d.ts files for package consumers

## [0.2.0] - 2020-06-23
### Added
- TypeScript Support

## [0.1.2] - 2020-06-20
### Added
- Simple server launcher
- Server can accept mods to load
- Server has a variety of options (silent, serverDir, mods, bots, steamAPIKey)
- SteamApiKey integration is WIP
- Bots are untested

[Unreleased]: https://github.com/brisberg/screeps-test-server/compare/v0.2.1...HEAD
[0.2.1]: https://github.com/brisberg/screeps-test-server/releases/tag/v0.2.0...v0.2.1
[0.2.0]: https://github.com/brisberg/screeps-test-server/releases/tag/v0.1.2...v0.2.0
[0.1.2]: https://github.com/brisberg/screeps-test-server/releases/tag/v0.1.2


test.spec.js
const del = require('del');
const fs = require('fs');
const path = require('path');
const ScreepsTestServer = require('./test-server');

describe('Screeps Test Server', () => {
  let server;

  afterEach(async () => {
    await server.stop();
  })

  it('should launch a test server', async () => {
    server = new ScreepsTestServer();
    await server.start();
    const {db, env, pubsub} = server;

    // Queries against live servers succeed
    expect(await env.get(env.keys.GAMETIME)).toEqual(2);

    await server.stop();

    const spy = jest.spyOn(console, 'error').mockImplementation();
    // Queries after server shutdown will error
    db['rooms'].find();
    expect(console.error).toHaveBeenCalled();
    spy.mockRestore();
  });

  describe(`'serverDir' server option`, () => {
    it('should initialize the default environment path', async () => {
      server = new ScreepsTestServer();
      await server.start();

      const envPath = path.join(process.cwd(), 'server');
      expect(fs.existsSync(path.join(envPath, '.screepsrc'))).toBeTruthy();
      expect(fs.existsSync(path.join(envPath, 'db.json'))).toBeTruthy();
      expect(fs.existsSync(path.join(envPath, 'mods.json'))).toBeTruthy();
      expect(fs.existsSync(path.join(envPath, 'steam_appid.txt'))).toBeTruthy();
    });

    it('should initialize the given server environment path', async () => {
      server = new ScreepsTestServer({serverDir: 'testEnv'});
      await server.start();

      const envPath = path.join(process.cwd(), 'testEnv');
      expect(fs.existsSync(path.join(envPath, '.screepsrc'))).toBeTruthy();
      expect(fs.existsSync(path.join(envPath, 'db.json'))).toBeTruthy();
      expect(fs.existsSync(path.join(envPath, 'mods.json'))).toBeTruthy();
      expect(fs.existsSync(path.join(envPath, 'steam_appid.txt'))).toBeTruthy();

      // Cleanup
      await del(envPath);
    });
  });

  it.todo(`should load server mods requested in options`);

  it.todo(`should load bot scripts requested in options`);

  describe.skip(`'silent' server option`, () => {
    let spy;

    beforeEach(() => {
      spy = jest.spyOn(process.stdout, 'write');
    })

    afterEach(() => {
      spy.mockRestore();
    })

    it(`should produce no console output with 'silent': true`, async () => {
      server = new ScreepsTestServer({silent: true});

      await server.start();

      expect(process.stdout.write).not.toHaveBeenCalled();
    });

    it(`should forward output to console with 'silent: false'`, async () => {
      server = new ScreepsTestServer({silent: false});

      await server.start();

      expect(process.stdout.write).toHaveBeenCalled();
    });
  });
});


src/mods.spec.ts
import fs from 'fs';
import path from 'path';

import ScreepsTestServer from './test-server';

describe('Screeps Test Server (with mods)', () => {
  let server: ScreepsTestServer;

  beforeEach(async () => {
    server = new ScreepsTestServer({
      mods: ['../lib/testing/simpleMod/index.js'],
    });
    await server.start();
  });

  afterEach(async () => {
    await server.stop();
  });

  it('should modify mods.json with given mod paths', async () => {
    const modsPath = path.join(process.cwd(), 'server', 'mods.json');
    const mods = JSON.parse(fs.readFileSync(modsPath, 'utf-8'));

    expect(mods.mods).toEqual(['../lib/testing/simpleMod/index.js']);
  });

  it('should launch test server with specified mod enabled', async (done) => {
    // See testing/simpleMod/index.js for implementation
    const {pubsub} = server;

    pubsub.subscribe('simplemod:out', (message: string) => {
      expect(message).toEqual('foobar_out');
      done();
    });

    pubsub.publish('simplemod:in', 'foobar');
  });
});


src/index.spec.ts
import {name} from './index';

describe('Package Index', () => {
  it('should export a constant', () => {
    expect(name).toBe('typescript-pkg');
  });
});


jest.config.js
module.exports = {
  // -- Default JestSettings (brisberg/typescript-pkg)

  // An array of glob patterns indicating a set of files for which coverage
  // information should be collected
  collectCoverageFrom: ['src/**/*.{js,ts}'],

  // An array of regexp pattern strings used to skip coverage collection
  coveragePathIgnorePatterns: ['/lib/', '\\.d\\.ts$'],

  // An array of directory names to be searched recursively up from the
  // requiring module's location
  moduleDirectories: ['node_modules', 'src'],

  // A preset that is used as a base for Jest's configuration
  preset: 'ts-jest',

  // The test environment that will be used for testing
  testEnvironment: 'node',

  // The glob patterns Jest uses to detect test files
  testMatch: ['**/*.spec.ts', '!**/node_modules/**', '!**/lib/**'],

  // -- Add package specific configurations below here
};


test-server.js
const fs = require('fs');
const path = require('path');
const cp = require('child_process');
const common = require('@brisberg/common');

// Arbitrary port number. Must match port hardcoded in .screepsrc
process.env.STORAGE_PORT = 24837;
// process.env.STORAGE_HOST = 'localhost';

/**
 * Screeps Test Server
 *
 * Handles parsing server options, managing screeps server processes, launching
 * separate common/storage process.
 */
class ScreepsTestServer {
  constructor(opts = {}) {
    this._connected = false;
    this._serverProcess = undefined;

    // Options
    this.silent = opts.silent || true;
    this.steamApiKey = opts.steamApiKey || '';
    this.envDir = opts.serverDir || 'server';
    this.mods = opts.mods || [];
    this.bots = opts.bots || [];
  }

  get db() {
    return common.storage.db;
  }

  get env() {
    return common.storage.env;
  }

  get pubsub() {
    return common.storage.pubsub;
  }

  get connected() {
    return this._connected;
  }

  /**
   * Launches a new Screeps Server, overwrites the database file, and connects
   * to the storage process.
   *
   * Stores the database, and environment, and pubsub handles.
   */
  async start() {
    // Copy server files into test environment
    const ASSETS_PATH = path.join(__dirname, 'assets');
    const TEST_ENV_PATH = path.join(process.cwd(), this.envDir);
    const SERVER_FILES = ['.screepsrc', 'db.json', 'steam_appid.txt'];
    fs.mkdirSync(TEST_ENV_PATH, {recursive: true});
    SERVER_FILES.forEach((fileName) => {
      fs.copyFileSync(
          path.join(ASSETS_PATH, fileName),
          path.join(TEST_ENV_PATH, fileName),
      );
    });
    // Write out mods.json file
    const modsData = {mods: this.mods, bots: this.bots};
    const modsJson = JSON.stringify(modsData, undefined, 2);
    fs.writeFileSync(path.join(TEST_ENV_PATH, 'mods.json'), modsJson);

    // Launch Server
    this._serverProcess = forkServerProcess(
        this.envDir,
        this.silent,
        this.steamApiKey,
    );
    // Wait for storage to initialize
    await new Promise((resolve) => setTimeout(resolve, 1000));
    await common.storage._connect();
    // await common.storage.resetAllData();
    this._connected = true;
  }

  /**
   * Tear down and cleanup the screeps process.
   */
  async stop() {
    if (!this.connected || !this._serverProcess) {
      return;
    }

    this._serverProcess.kill();
    // common.storage._connected = false;
    this._serverProcess = undefined;
    this._connected = false;

    // Wait for process to die
    return new Promise((resolve) => setTimeout(resolve, 50));
  };
}
module.exports = ScreepsTestServer;

// Forks a full Screeps Server process and returns the handle
function forkServerProcess(serverDir, silent = true, steamApiKey = '') {
  const execPath = path.resolve(
      path.dirname(require.resolve('@screeps/launcher')),
      '../bin/screeps.js',
  );
  return cp.fork(
      path.resolve(execPath),
      ['start', '--steam_api_key', steamApiKey],
      {
        cwd: path.join(process.cwd(), serverDir),
        silent,
        stdio: silent ? 'ignore' : 'inherit',
      },
  );
};


src/test-server.ts
import cp from 'child_process';
import fs from 'fs';
import path from 'path';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const common = require('@brisberg/common');

// Arbitrary port number. Must match port hardcoded in .screepsrc
process.env.STORAGE_PORT = '24837';
// process.env.STORAGE_HOST = 'localhost';

export interface ScreepsTestServerOptions {
  silent?: boolean;
  steamApiKey?: string;
  serverDir?: string;
  mods?: string[];
  bots?: string[];
}

/**
 * Screeps Test Server
 *
 * Handles parsing server options, managing screeps server processes, launching
 * separate common/storage process.
 */
export default class ScreepsTestServer {
  private _connected = false;
  private serverProcess?: cp.ChildProcess = undefined;  // Server process handle
  private silent = true;
  private steamApiKey = '';
  private envDir = 'server';  // Directory name of test environment
  private mods: string[] = [];
  private bots: string[] = [];

  constructor(opts: ScreepsTestServerOptions = {}) {
    // Options
    this.silent = opts.silent || true;
    this.steamApiKey = opts.steamApiKey || '';
    this.envDir = opts.serverDir || 'server';
    this.mods = opts.mods || [];
    this.bots = opts.bots || [];
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  get db(): any {
    return common.storage.db;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  get env(): any {
    return common.storage.env;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  get pubsub(): any {
    return common.storage.pubsub;
  }

  get connected(): boolean {
    return this._connected;
  }

  /**
   * Launches a new Screeps Server, overwrites the database file, and connects
   * to the storage process.
   *
   * Stores the database, and environment, and pubsub handles.
   */
  async start(): Promise<void> {
    // Copy server files into test environment
    const ASSETS_PATH = path.join(__dirname, '..', 'assets');
    const TEST_ENV_PATH = path.join(process.cwd(), this.envDir);
    const SERVER_FILES = ['.screepsrc', 'db.json', 'steam_appid.txt'];
    fs.mkdirSync(TEST_ENV_PATH, {recursive: true});
    SERVER_FILES.forEach((fileName) => {
      fs.copyFileSync(
          path.join(ASSETS_PATH, fileName),
          path.join(TEST_ENV_PATH, fileName),
      );
    });
    // Write out mods.json file
    const modsData = {mods: this.mods, bots: this.bots};
    const modsJson = JSON.stringify(modsData, undefined, 2);
    fs.writeFileSync(path.join(TEST_ENV_PATH, 'mods.json'), modsJson);

    // Launch Server
    this.serverProcess = forkServerProcess(
        this.envDir,
        this.silent,
        this.steamApiKey,
    );
    // Wait for storage to initialize
    await new Promise((resolve) => setTimeout(resolve, 1000));
    await common.storage._connect();
    // await common.storage.resetAllData();
    this._connected = true;
  }

  /**
   * Tear down and cleanup the screeps process.
   */
  async stop(): Promise<void> {
    if (!this.connected || !this.serverProcess) {
      return;
    }

    this.serverProcess.kill();
    // common.storage._connected = false;
    this.serverProcess = undefined;
    this._connected = false;

    // Wait for process to die
    return new Promise((resolve) => setTimeout(resolve, 50));
  }
}

// Forks a full Screeps Server process and returns the handle
function forkServerProcess(
    serverDir: string, silent = true, steamApiKey = 'none'): cp.ChildProcess {
  const execPath = path.resolve(
      path.dirname(require.resolve('@screeps/launcher')),
      '../bin/screeps.js',
  );
  return cp.fork(
      path.resolve(execPath),
      ['start', '--steam_api_key', steamApiKey],
      {
        cwd: path.join(process.cwd(), serverDir),
        silent,
        stdio: silent ? 'ignore' : 'inherit',
      },
  );
}
